subprojects/workers/src/main/java/org/gradle/workers/WorkerExecutor.java

```diff
@@ -20,26 +20,38 @@
 import org.gradle.api.Incubating;
 
 /**
- * Worker Executor.
+ * Allows work to be submitted for asynchronous execution.  Work should be submitted with a {@link Runnable} class
+ * representing the implementation of the unit of work and an action to configure the unit of work (via {@link WorkerConfiguration}).
+ *
+ * <pre>
+ *      workerExecutor.submit(RunnableWorkImpl.class) { WorkerConfiguration conf ->
+ *          // Set the isolation mode for the worker
+ *          conf.isolationMode = IsolationMode.NONE
+ *
+ *          // Set up the constructor parameters for the unit of work
+ *          conf.params = [ "foo", file('bar') ]
+ *      }
+ * </pre>
  *
  * @since 3.5
  */
 @Incubating
 public interface WorkerExecutor {
     /**
-     * Submits a piece of work to be executed.
+     * Submits a piece of work to be executed asynchronously.
      *
      * Execution of the work may begin immediately.
      *
-     * Forked work will execute in an idle daemon that meets the requirements set on this builder.  If no
-     * idle daemons are available, a new daemon will be started.  Any errors will be thrown from  {@link #await()}.
-     *
-     * In the event that an error is thrown while submitting work, all uncompleted work will be canceled.
+     * Work configured with {@link IsolationMode#PROCESS} will execute in an idle daemon that meets the requirements set
+     * in the {@link WorkerConfiguration}.  If no idle daemons are available, a new daemon will be started.  Any errors
+     * will be thrown from {@link #await()} or from the surrounding task action if {@link #await()} is not used.
      */
     void submit(Class<? extends Runnable> actionClass, Action<? super WorkerConfiguration> configAction);
 
     /**
-     * Blocks until all work associated with the current build operation is complete.
+     * Blocks until all work associated with the current build operation is complete.  Note that when using this method inside
+     * a task action, it will block completion of the task action until all submitted work is complete.  This means that other
+     * tasks from the same project cannot be run in parallel while the task action is still executing.
```

Comment:
Yes, it is.  There are two main reasons why this is the case, both related to safety.  The first is that while running a task action, the task should have exclusive access to the state of the Project.  If we surrender control to another task, it could potentially change that state in the middle of the task action, so we set the boundary for this access around the task action and don't release it even if we call `await()`.  Second, and perhaps more compelling, is that allowing another task to run while we are still in the task action can more easily lead to deadlocks.  Imagine two tasks of type Foo, the first (foo1) starts and acquires a lock inside the task action, then submits work and waits for it.  This releases access to the Project and allows the other Foo task (foo2) to run, which immediately blocks trying to acquire the same lock as foo1.  Now, you have a deadlock with foo1 holding the lock, but can't get control back to release it, while 2 is waiting on the lock, but won't give up control.

So, to solve the use case that you suggest - chaining work together - we want to allow units of work to inject the WorkerExecutor and be able to submit units of work themselves.  So in your example, the task action would submit one unit of work and release Project access so another task can start.  That unit of work would then submit your three units of work, wait, and then submit the fourth.  Since it all happens outside of the task action, we don't have the same concerns around controlling access to the Project.  This feature (service injection in workers) is one of the next ones we'll tackle.